both/access_control.js

```diff
@@ -231,101 +229,75 @@ if(Meteor.isServer) {
         'createdAt',
       ];
 
-      if(_.difference(fields, allowedFields).length > 0) {
+      return _onlyAllowedFields(fields, allowedFields);
+    },
+    fetch: [ 'competitionId' ],
+  });
+
+  RoundProgresses.allow({
+    update: function(userId, progress, fields, modifier) {
+      if(getCannotManageCompetitionReason(userId, progress.competitionId)) {
         return false;
       }
-      return true;
+      var allowedFields = [
+        'done',
+        'total',
+      ];
+      return _onlyAllowedFields(fields, allowedFields);
     },
-    fetch: [ 'competitionId' ],
   });
 
   Results.allow({
     update: function(userId, result, fields, modifier) {
-      var competition = Competitions.findOne({ _id: result.competitionId }, { fields: { _id: 1 } });
-      if(getCannotManageCompetitionReason(userId, competition._id)) {
+      if(getCannotManageCompetitionReason(userId, result.competitionId)) {
         return false;
       }
-
       var allowedFields = [
         'solves',
 
         'updatedAt',
         'createdAt',
       ];
-
-      if(_.difference(fields, allowedFields).length > 0) {
-        return false;
-      }
-      return true;
+      return _onlyAllowedFields(fields, allowedFields);
     },
     fetch: [ 'competitionId' ],
   });
 
-  var allowedOwnRegistrationFields = [
-    'userId',
-    'competitionId',
-    'uniqueName',
-    'registeredEvents',
-    'guestCount',
-    'comments',
-
-    'updatedAt',
-    'createdAt',
-  ];
   Registrations.allow({
     insert: function(userId, registration) {
-      if(!getCannotManageCompetitionReason(userId, registration.competitionId)) {
-        // If you're allowed to manage the competition, then you can
-        // add any registration.
-        return true;
-      }
-      if(getCannotRegisterReasons(registration.competitionId)) {
-        return false;
-      }
-      // can only edit entries with their user id
-      if(registration.userId != userId) {
-        return false;
-      }
-
-      if(_.difference(fields, allowedOwnRegistrationFields).length > 0) {
-        return false;
-      }
-      return true;
+      return Registrations.userIsAllowed(userId, registration);
```

Comment:
Some mutual confusion here.

1. This *is* the #209 bug.

2. When I saw that insert() doesn't get the field names I decided that enforcing what fields was not needed. And that may be the intention of the Meteor developers, but I see now that there are 8 fields that are not in the allowed list, and I mindlessly threw that functionality out.

I'll put that back in, but it doesn't seem like the right long term implementation. For one thing, accepting userId and competitionId from an untrustable client is really problematic. Something like the addRound() Method would make more sense, and also make the app consistent.

Philosophical design points:

3. I think listing field names in this code is a pretty bad DRY violation. We already define the field names in the SimpleSchemas, and this info should be custom fields there.

4. Ultimately, if one set of fields needs to be treated differently than another set, that is probably a good indication they belong in separate classes/collections. Payment information definitely sounds like a separate class to me.